### PR TITLE
ZPM support

### PR DESCRIPTION
**Pull request by daimor**

### CODE REVIEW

**Comment by ZitRos** [comment]
Thanks, Dmitry! I can look through compatibility on different installations once I have some time.

But anyway, I see a couple of cases not implemented here:

1. How does mapping to %All works here? Does it use projections anyway?
2. As well as mapping to all non-%All namespaces like SAMPLES and DOCBOOK?
3. There should be [WebTerminal](https://github.com/intersystems-community/webterminal/blob/4d6dc00c9e1c166a3f9b63092dd722812c4b98d5/src/cls/WebTerminal/Installer.cls#L131) role set up, as well as [WebTerminal Resource](https://github.com/intersystems-community/webterminal/blob/4d6dc00c9e1c166a3f9b63092dd722812c4b98d5/src/cls/WebTerminal/Installer.cls#L126)
4. WebTerminal [reports installation status](https://github.com/intersystems-community/webterminal/blob/4d6dc00c9e1c166a3f9b63092dd722812c4b98d5/src/cls/WebTerminal/Installer.cls#L63). Can this work in ZPM?

Still, I beg you, please, please, allow projections run in ZPM. Why not? Did you intentionally turn them off? You won't be able to cover everything with the install manifest; ISC products are just designed differently from the very beginning.

What you can do though, and what will be the most elegant thing is

```
RunProjections="1"
```

In the manifest :)

**Comment by daimor** [comment]
I can't just turn it off. I just did not put it in the package, that's it.
There are some issues with mapping, we don't have it, yet implemented in zpm. And another issue is, that IRIS Community Edition, does not allow mapping at all.
I think the best solution for ZPM will be something like [npm scripts](https://docs.npmjs.com/misc/scripts), so you would be able to call some code by some callbacks. Where you would be able to call installations reports, for instance.

Setting roles or any security should be as part of the manifest.

So, I'm sure every rollable change should be configured by manifest. Only in this case, I can be sure that instance will be better cleared after uninstall.

**Comment by ZitRos** [comment]
> I can't just turn it off. I just did not put it in the package, that's it.
Say what?

Well, whatever we do in ZPM, I don't think that I will remove projections. Simply because of compatibility issues. Furthermore, supporting two installers and scripts is waaaay too complex...

Hence,
1. I'll maintain projections for 3 of my projects anyway
2. I'll need to maintain ZPM as well, I guess for newest ISC product versions

My opinion is still that `RunProjections="1"` is the best thing we can come up with.

> I think the best solution for ZPM will be something like npm scripts, so you would be able to call some code by some callbacks.

**Because if you allow scripts, you won't be ever able to**

> be sure that instance will be better cleared after uninstall.

**Comment by daimor** [comment]
I don't see any easy possibilities yet, for `RunProjections="1"`
As you know projection runs during compile, and we don't have any flags to stop it.

**Comment by ZitRos** [comment]
Wait, so they do run, don't they?

**Comment by daimor** [comment]
At this time, `WebTerminal.Installer` is just not in the package, if we include this class, the projection will work.

**Comment by ZitRos** [comment]
> At this time, WebTerminal.Installer is just not in the package, if we include this class, the projection will work.

🙏

File: gulpfile.babel.js [hunk 1]
```diff
@@ -120,22 +121,8 @@ gulp.task("css", ["scss", "copy-css-themes"], function (cb) {
...
-gulp.task("pre-cls", ["js", "js", "html", "css", "readme"], () => {
+gulp.task("cls", ["js", "js", "html", "css", "readme"], () => {
     return gulp.src([`${ source }/cls/**/*.cls`])
-        .pipe(rename((f) => {
-            f.basename = `${ f.dirname === "." ? "" : f.dirname + "." }${
-                f.basename
-            }`;
-            f.dirname = ".";
-            if (f.basename !== INSTALLER_CLASS_NAME)
-                context.context.compileAfter +=
```
Comment (ZitRos) [inline gulpfile.babel.js]: Getting back to this MR, @daimor, why did you remove `compileAfter` context variable?

Comment (daimor) [reply gulpfile.babel.js]: it is not just removing `compileAfter`, I have deleted task `pre-cls` completely, which was just about renaming static files. This not needed anymore, due to other my changes: includeBase and changed paths in `StaticContent.cls`

Comment (ZitRos) [reply gulpfile.babel.js]: I've just realized that you've excluded Installer class from the installation "pipeline" at all... @daimor, I can't merge this 😶

**Comment by ZitRos** [comment]
@daimor, to fix this thing I need to understand what ZPM build pipeline should look like:

1. What to output (I see `<SourcesRoot>build</SourcesRoot>` in `module.xml`, should we also create a `build_zpm` build, without Installer)?
2. How ZPM module is built and published. Do we even need `build_zpm` here, or is it manual for now?
3. I've just published WebTerminal v4.9.0. What is the process of publishing it to ZPM? Any docs on this?

If we just can put `module.xml` in the root of this repository this is fine only if all other jobs are manual. But I don't understand what your PR for. Thanks!

**Comment by daimor** [comment]
Sorry, was very busy last time. So, 
1. `build` is output folder for your gulp, and then it used as a source of code in zpm, only this folder will appear in a final published package.
2. build zpm package is manual by now, yet. We don't have any entrypoints from OS side.
3. at the moment publishing packages not yet in public, not everything is finished on that side. I can do it, when I will find time.
my ideas was even to use GitHub Actions for that process, but it was not done, yet.
you can look at tgz file, by url https://pm.community.intersystems.com/packages/webterminal
and look how your package looks inside.

**Comment by ZitRos** [comment]
@daimor, thanks!

Latest [Github Package Registry](https://github.com/features/package-registry) and [GitHub actions](https://github.com/features/actions) are definitely great things for ZPM.

As for the current PR contents, `build` directory in this repository has a very specific meaning. To release a new WebTerminal version I run `./import.bat`, which does create `build` directory, builds a project, then imports it into local Ensemble instance (I use 2017.1) and then export XML from there. **And your PR just makes this pipeline broken**.

What I would suggest is, because of the process of publishing things to ZPM is currently a manual process, we **just leave only `module.xml` file in the root of this repo**, without altering a general pipeline (by the way I cannot make it as a part of an automated CI pipeline, because exported files from 2019.* public Docker image seems to not import in previous versions). In other words, **accepting only `module.xml` file from this PR**.

Does this make sense? Can we close this PR then?

**Comment by ZitRos** [comment]
Alternatively, if we could create an automated CI pipeline for publishing WebTerminal to ZPM, as well as its XML, I would suggest creating a separate task for it, like the output to `build_zpm` or so.

**Comment by daimor** [comment]
It is not a big problem to make CI, for previous versions and for latest version, and build xml sources for older version from latest versions.
I would keep this PR, yet, until I will find the time, to check your concerns.

**Comment by ZitRos** [comment]
OK 👌

> It is not a big problem to ... build xml sources for older version from latest versions.

At least I tried that, and it seems to be broken, unfortunately (because of new tags and properties in XML file introduced in 2019), as well as some other incompatibilities. They can be fixed out of IRIS instance (by simply replacing stuff in XML), but this would be a temporary crutch until the next version of IRIS kicks in (though we can live with 2019 image forever).

**Comment by daimor** [comment]
Did you try to export sources with flag /[exportversion](https://docs.intersystems.com/irislatest/csp/docbook/DocBook.UI.Page.cls?KEY=RCOS_vsystem)=2010.2
it should work even from IRIS to Cache, not back, yet.
IRIS had some issues in downgrading XML, but it should be fixed in 2019.3

**Comment by ZitRos** [comment]
> Did you try to export sources with flag /exportversion=2010.2

Sounds really appealing, thanks! I'll definitely try that.

**Comment by daimor** [comment]
Changed a bit, can it be merged now?
Just only don't know how to deal with version in module.xml
